waspc/cli/Wasp/Cli/Command/Watch.hs

```diff
@@ -29,25 +31,37 @@ import qualified Wasp.Lib
 
 -- | Forever listens for any file changes in waspProjectDir, and if there is a change,
 --   compiles Wasp source files in waspProjectDir and regenerates files in outDir.
+--   It will defer recompilation until no change was detected in the last second.
 watch :: Path' Abs (Dir Common.WaspProjectDir) -> Path' Abs (Dir Wasp.Lib.ProjectRootDir) -> IO ()
 watch waspProjectDir outDir = FSN.withManager $ \mgr -> do
   currentTime <- getCurrentTime
   chan <- newChan
   _ <- FSN.watchDirChan mgr (SP.fromAbsDir waspProjectDir) eventFilter chan
   _ <- FSN.watchTreeChan mgr (SP.fromAbsDir $ waspProjectDir </> Common.extCodeDirInWaspProjectDir) eventFilter chan
-  listenForEvents chan currentTime
+  listenForEvents chan currentTime False
   where
-    listenForEvents :: Chan FSN.Event -> UTCTime -> IO ()
-    listenForEvents chan lastCompileTime = do
-      event <- readChan chan
-      let eventTime = FSN.eventTime event
-      if eventTime < lastCompileTime
-        then -- If event happened before last compilation started, skip it.
-          listenForEvents chan lastCompileTime
-        else do
+    listenForEvents :: Chan FSN.Event -> UTCTime -> Bool -> IO ()
+    listenForEvents chan lastCompileTime recompileAfterPreemptibleDelay = do
+      let microsecondsPerSecond = 1000000
+      res <-
+        if recompileAfterPreemptibleDelay
+          then -- Recompile after 1s, unless we get another file change before then.
+            race (readChan chan) (threadDelay $ 1 * microsecondsPerSecond)
+          else -- Just wait for next change.
+            Left <$> readChan chan
+      case res of
+        Left event -> do
```

Comment:
Usage of Left and Right -> sure, it looks ok! This is how `race` uses them anyway, so I think you did all good here. `Either` is often used for storing error in `Left` and value in `Right`, main reason being that `Either` is functor in the `Right` value, but here we are using it for different purpose and that is ok.

Reply:
Aha, got what you mean by swapping Left and Right -> again, I think it doesn't matter, it is up to you completely. Only thing you might want to keep in mind is this Functor thing, but in this case it really doesn't matter. It makes a bit of sense to me to have Right as wanted behaviour, while Left is kind of unwanted -> and since we want delay to succeed, it makes just a bit more sense to me to keep it in Right, as you did initially. But again, not really important here.

Reply:
Nice, thanks for confirming that! I did swap in a cleanup refactor out of "just in case" :D , but good to know it was not strictly necessary!